scripts/do-release.py

```diff
@@ -0,0 +1,188 @@
+
+
+class BaseCredentials:
+    """Base credential class."""
+
+    @abstractproperty
+    def pypi_username(self) -> str:
+        """Get PYPI username."""
+        pass
+
+    @abstractproperty
+    def pypi_password(self) -> str:
+        """Get PYPI password."""
+        pass
+
+
+class EnvCredentials(BaseCredentials):
+    """Credentials from env variables."""
+
+    @property
+    def pypi_username(self) -> str:
+        """Get PYPI username."""
+        return os.environ.get("PYPI_USERNAME") or ""
+
+    @property
+    def pypi_password(self) -> str:
+        """Get PYPI password."""
+        return os.environ.get("PYPI_PASSWORD") or ""
+
+
+class ReleaseTool:
+    """Release helper tool."""
+
+    def __init__(self, credentials: BaseCredentials) -> None:
+        """Init release tool instance."""
+        self._credentials = credentials
+
+    def get_the_latest_release_version(self) -> Version:
+        """Get release version from gihtub tags."""
+        text = subprocess.check_output(
+            "git ls-remote --tags origin", shell=True, text=True
+        )
+        tags = [i.split("\t")[1].strip() for i in text.splitlines()]
+        tags = [
+            i for i in tags if i.startswith("refs/tags/v") and not i.endswith("^{}")
+        ]
+        versions = [i.replace("refs/tags/v", "") for i in tags]
+        return Version(versions[-1])
+
+    def get_current_version(self) -> Version:
+        """Get current code version."""
+        text = (ROOT / "pyproject.toml").read_text()
+        version = tomli.loads(text)["tool"]["poetry"]["version"]
+        return Version(version)
+
+    def do_we_need_to_release(self) -> bool:
+        """Check is code version is newer than on github."""
+        current_version = self.get_current_version()
+        released_version = self.get_the_latest_release_version()
+        return current_version > released_version
+
+    def parse_history(self) -> Tuple[List, Dict]:
+        """Parse HISTORY.md."""
+        text = (ROOT / "HISTORY.md").read_text()
+        versions = [Version(i) for i in re.findall("## ([^ \n]+)", text)]
+        history = {
+            Version(i[0]): i[2]
+            for i in re.findall(
+                r"^## ([^ ]+)\n?([^\n]+\n)?\n([^#]+)", text, re.DOTALL + re.M
+            )
+        }
+        return versions, history
+
+    def make_tag(self, current_version: Version) -> None:
+        """Make git tag."""
+        subprocess.check_call(
+            f"git tag v{current_version} -m 'Release {current_version}'", shell=True
+        )
+
+    def push_tag(self, current_version) -> None:
+        """Push tag to github."""
+        subprocess.check_call(f"git push origin v{current_version}", shell=True)
+
+    def make_release(self, current_version: Version, release_history: str) -> None:
+        """Make release on Github."""
+        subprocess.check_call(
+            f"""gh release create v{current_version} --title "v{current_version}" --notes "{release_history}" """,
```

Comment:
Can use the GH's autogenerated changelog instead of what's in the HISTORY.md?

Reply:
this requires good changelog on github, probably need to agree on merge commits format
for the moment it's better to generate manually